### PR TITLE
[New] Display points using clustering feature reduction

### PR DESCRIPTION
**Pull request by des12437**
## Description

This PR implements `Display points using clustering feature reduction` in `Visualization` category.
URL to README: [URL](https://github.com/Esri/arcgis-maps-sdk-swift-samples/tree/des12437/Display-points-using-clustering-feature-reduction/Shared/Samples/Display%20points%20using%20clustering%20feature%20reduction)

## Linked Issue(s)

- `swift/issues/4405`

## How To Test
- Tap on clustered geo element to view a Popup of aggregated information and summary statistics for the cluster. 
- Toggle the Feature clustering switch on and off to enable and disable clustering on the feature layer respectively.

## Screenshots

display-points-using-clustering-feature-reduction

## To Discuss

- Map View does not fill entire view on iPhone at the top of the sample. Should the Map View have aspectFill applied to it's layout?
- Should the toggle switch have a different layout? The automatic layout adds padding between the label and switch.
- Dateline shiproom issue runtime/orion/issues/1082

### CODE REVIEW

File: Shared/Samples/Display points using clustering feature reduction/DisplayPointsUsingClusteringFeatureReductionView.swift [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import ArcGIS
+import SwiftUI
+import ArcGISToolkit
+
+struct DisplayPointsUsingClusteringFeatureReductionView: View {
+    /// A map of global power plants.
+    @State private var map = {
+        let portalItem = PortalItem(
+            portal: .arcGISOnline(connection: .anonymous),
+            id: Item.ID(rawValue: "8916d50c44c746c1aafae001552bad23")!
```
Comment (yo1995) [inline Shared/Samples/Display points using clustering feature reduction/DisplayPointsUsingClusteringFeatureReductionView.swift:24]:
```suggestion
            id: PortalItem.ID("8916d50c44c746c1aafae001552bad23")!
```

File: Shared/Samples/Display points using clustering feature reduction/DisplayPointsUsingClusteringFeatureReductionView.swift [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    @State private var layer: FeatureLayer?
+    
+    /// The screen point to perform an identify operation.
+    @State private var identifyScreenPoint: CGPoint?
+    
+    /// The popup to be shown as the result of the layer identify operation.
+    @State private var popup: Popup?
+    
+    /// A Boolean value specifying whether the popup view should be shown or not.
+    @State private var showPopup = false
```
Comment (yo1995) [inline Shared/Samples/Display points using clustering feature reduction/DisplayPointsUsingClusteringFeatureReductionView.swift:39]: Should it be "showsPopup"?

File: Shared/Samples/Display points using clustering feature reduction/DisplayPointsUsingClusteringFeatureReductionView.swift [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    
+    /// A Boolean value specifying whether the layer's feature reduction is shown.
+    @State var showsFeatureReduction = true
```
Comment (yo1995) [inline Shared/Samples/Display points using clustering feature reduction/DisplayPointsUsingClusteringFeatureReductionView.swift:42]:
```suggestion
    @State private var showsFeatureReduction = true
```

File: Shared/Samples/Display points using clustering feature reduction/DisplayPointsUsingClusteringFeatureReductionView.swift [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    
+    /// A Boolean value indicating whether to show an error alert.
+    @State var isShowingAlert = false
```
Comment (yo1995) [inline Shared/Samples/Display points using clustering feature reduction/DisplayPointsUsingClusteringFeatureReductionView.swift:45]:
```suggestion
    @State private var isShowingAlert = false
```

File: Shared/Samples/Display points using clustering feature reduction/DisplayPointsUsingClusteringFeatureReductionView.swift [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    
+    /// The error shown in the error alert.
+    @State var error: Error? {
```
Comment (yo1995) [inline Shared/Samples/Display points using clustering feature reduction/DisplayPointsUsingClusteringFeatureReductionView.swift:48]:
```suggestion
    @State private var error: Error? {
```

File: Shared/Samples/Display points using clustering feature reduction/DisplayPointsUsingClusteringFeatureReductionView.swift [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    
+    var body: some View {
+        MapViewReader { proxy in
+            MapView(map: map)
+                .onSingleTapGesture { screenPoint, _ in
+                    identifyScreenPoint = screenPoint
+                }
+                .task(id: identifyScreenPoint) {
+                    guard let identifyScreenPoint = identifyScreenPoint,
+                          let layer = self.layer,
```
Comment (yo1995) [inline Shared/Samples/Display points using clustering feature reduction/DisplayPointsUsingClusteringFeatureReductionView.swift:60]:
```suggestion
                          let layer,
```

File: Shared/Samples/Display points using clustering feature reduction/DisplayPointsUsingClusteringFeatureReductionView.swift [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    }
+    
+    var body: some View {
+        MapViewReader { proxy in
+            MapView(map: map)
+                .onSingleTapGesture { screenPoint, _ in
+                    identifyScreenPoint = screenPoint
+                }
+                .task(id: identifyScreenPoint) {
+                    guard let identifyScreenPoint = identifyScreenPoint,
```
Comment (yo1995) [inline Shared/Samples/Display points using clustering feature reduction/DisplayPointsUsingClusteringFeatureReductionView.swift:59]:
```suggestion
                    guard let identifyScreenPoint,
```

File: Shared/Samples/Display points using clustering feature reduction/DisplayPointsUsingClusteringFeatureReductionView.swift [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    @State private var map = {
+        let portalItem = PortalItem(
+            portal: .arcGISOnline(connection: .anonymous),
+            id: Item.ID(rawValue: "8916d50c44c746c1aafae001552bad23")!
+        )
+        return Map(item: portalItem)
+    }()
+    
+    /// The power plants feature layer for querying.
+    @State private var layer: FeatureLayer?
```
Comment (yo1995) [inline Shared/Samples/Display points using clustering feature reduction/DisplayPointsUsingClusteringFeatureReductionView.swift:30]: The layer can be a computed property from the map, no need to be a state.

File: Shared/Samples/Display points using clustering feature reduction/README.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+Pan and zoom the map to view how clustering is dynamically updated. Toggle clustering off to view the original point features that make up the clustered elements. When clustering is toggled on, you can tap on a clustered geoelement to view aggregated information and summary statistics for that cluster. When clustering is toggled off and you tap on the original feature you get access to information about individual power plant features.
+
+## How it works
+
+1. Create a map from a web map `PortalItem`.
+2. Get the cluster enabled layer from the map's operational layers.
+3. Get the `FeatureReduction` from the feature layer and set `isEnabled` to enable or disable clustering on the feature layer.
+4. Use the `onSingleTapGesture` modifier to listen for tap events on the map view.
+5. Identify tapped features on the map using `identify(on layer:screenPoint:tolerance:)` on the feature layer and pass in the map screen point location.
```
Comment (yo1995) [inline Shared/Samples/Display points using clustering feature reduction/README.md:21]:
```suggestion
5. Identify tapped features on the map using `identify(on:screenPoint:tolerance:returnPopupsOnly:maximumResults:)` on the feature layer and pass in the map screen point location.
```

File: Shared/Samples/Display points using clustering feature reduction/README.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+* AggregateGeoElement
+* FeatureLayer
+* FeatureReduction
+* GeoElement
+* IdentifyLayerResult
+
+## About the data
+
+This sample uses a [web map](https://www.arcgis.com/home/item.html?id=8916d50c44c746c1aafae001552bad23) that displays the [Esri Global Power](https://www.arcgis.com/home/item.html?id=eb54b44c65b846cca12914b87b315169) Plants feature layer with feature reduction enabled. When enabled, the aggregate features symbology shows the color of the most common power plant type, and a size relative to the average plant capacity of the cluster.
```
Comment (yo1995) [inline Shared/Samples/Display points using clustering feature reduction/README.md:35]:
```suggestion
This sample uses a [web map](https://www.arcgis.com/home/item.html?id=8916d50c44c746c1aafae001552bad23) that displays the [Esri Global Power Plants](https://www.arcgis.com/home/item.html?id=eb54b44c65b846cca12914b87b315169) feature layer with feature reduction enabled. When enabled, the aggregate features symbology shows the color of the most common power plant type, and a size relative to the average plant capacity of the cluster.
```

**Review by yo1995 — COMMENTED** [review]
> Map View does not fill entire view on iPhone at the top of the sample. Should the Map View have aspectFill applied to it's layout?

Do you mean the initial extent of the web map? Maybe we can set a different initial viewpoint.

> Should the toggle switch have a different layout? The automatic layout adds padding between the label and switch.

Can use `ToolbarItemGroup` with `Spacer`

**Comment by des12437** [comment]
> > Map View does not fill entire view on iPhone at the top of the sample. Should the Map View have aspectFill applied to it's layout?
> 
> Do you mean the initial extent of the web map? Maybe we can set a different initial viewpoint.

Yes, however I'm not sure how to derive a good initial viewpoint for the [webmap](https://www.arcgis.com/home/item.html?id=8916d50c44c746c1aafae001552bad23).

> > Should the toggle switch have a different layout? The automatic layout adds padding between the label and switch.
> 
> Can use `ToolbarItemGroup` with `Spacer`

This is the result with that layout. I'm not sure if this looks better than the original layout.

**Review by yo1995 — APPROVED** [review]
I'm fine with the toggle layout. IMO the viewpoint can just be zoomed in a bit, such as `await proxy.setViewpointScale(1e7)`. If you want to reset it, it can be either around middle east where the oil clusters are the most prominent, or in the US to show the diversity of types of clusters. 

Redlands can always be used as a fallback viewpoint center location.

**Review by yo1995 — COMMENTED** [review]
When I test on an iPad, if I tap the same cluster for the second time after I dismiss the popup, say "Kalinn NPP", the app crashes. Do you see this as well?

**Comment by des12437** [comment]
> When I test on an iPad, if I tap the same cluster for the second time after I dismiss the popup, say "Kalinn NPP", the app crashes. Do you see this as well?

I don't see this on iPad. Does this look like the right way to reproduce the issue?

https://github.com/Esri/arcgis-maps-sdk-swift-samples/assets/117859673/e90ee6b8-b334-4dc6-a0d9-0550fbc0c97f

**Comment by yo1995** [comment]
> I don't see this on iPad. Does this look like the right way to reproduce the issue?

Yes, try a few more times to tap around the clusters. I got the following error

> "request for layout attributes for decoration view of kind _UICollectionViewListLayoutSectionBackgroundColorDecorationElementKind in section 0 when there are only 0 sections in the collection view"

It might be related to the popup, so not really a blocker for this PR.

<details><summary>stack trace</summary>

```txt
2023-08-09 09:26:39.446386-0700 Samples[84139:977760] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'request for layout attributes for decoration view of kind _UICollectionViewListLayoutSectionBackgroundColorDecorationElementKind in section 0 when there are only 0 sections in the collection view'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007ff80045478b __exceptionPreprocess + 242
	1   libobjc.A.dylib                     0x00007ff80004db73 objc_exception_throw + 48
	2   Foundation                          0x00007ff800b42f64 _userInfoForFileAndLine + 0
	3   UIKitCore                           0x000000010d26a877 -[UICollectionViewData layoutAttributesForDecorationViewOfKind:atIndexPath:nilAttributesArePermitted:] + 538
	4   UIKitCore                           0x000000010d26ea4e -[UICollectionViewData layoutAttributesForDecorationViewOfKind:atIndexPath:] + 17
	5   UIKitCore                           0x000000010d22e01c -[UICollectionView _dequeueReusableViewOfKind:withIdentifier:forIndexPath:viewCategory:] + 513
	6   UIKitCore                           0x000000010d2121f1 -[UICollectionView _createPreparedSupplementaryViewForElementOfKind:atIndexPath:layout:withLayoutAttributes:applyAttributes:] + 1040
	7   UIKitCore                           0x000000010d21c739 -[UICollectionView _createVisibleViewsForSingleCategoryAttributes:limitCreation:fadeForBoundsChange:] + 2144
	8   UIKitCore                           0x000000010d21cb0b -[UICollectionView _createVisibleViewsForAttributes:fadeForBoundsChange:notifyLayoutForVisibleCellsPass:] + 376
	9   UIKitCore                           0x000000010d21a7c3 -[UICollectionView _updateVisibleCellsNow:] + 2372
	10  UIKitCore                           0x000000010d221ad0 -[UICollectionView layoutSubviews] + 375
	11  SwiftUI                             0x00000001114b0a52 block_destroy_helper + 29479
	12  SwiftUI                             0x00000001114b0c14 block_destroy_helper + 29929
	13  UIKitCore                           0x000000010e364083 -[UIView(CALayerDelegate) layoutSublayersOfLayer:] + 2313
	14  QuartzCore                          0x00007ff808810d98 _ZN2CA5Layer16layout_if_neededEPNS_11TransactionE + 526
	15  QuartzCore                          0x00007ff80881c849 _ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE + 65
	16  QuartzCore                          0x00007ff80872c15e _ZN2CA7Context18commit_transactionEPNS_11TransactionEdPd + 624
	17  QuartzCore                          0x00007ff808762805 _ZN2CA11Transaction6commitEv + 723
	18  QuartzCore                          0x00007ff808763cb8 _ZN2CA11Transaction25flush_as_runloop_observerEb + 60
	19  CoreFoundation                      0x00007ff8003b1c12 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	20  CoreFoundation                      0x00007ff8003ac57f __CFRunLoopDoObservers + 515
	21  CoreFoundation                      0x00007ff8003acaa2 __CFRunLoopRun + 1121
	22  CoreFoundation                      0x00007ff8003ac264 CFRunLoopRunSpecific + 560
	23  GraphicsServices                    0x00007ff809b4024e GSEventRunModal + 139
	24  UIKitCore                           0x000000010dd297bf -[UIApplication _run] + 994
	25  UIKitCore                           0x000000010dd2e5de UIApplicationMain + 123
	26  SwiftUI                             0x0000000111f9f5ed __swift_memcpy93_8 + 12403
	27  SwiftUI                             0x0000000111f9f4a6 __swift_memcpy93_8 + 12076
	28  SwiftUI                             0x00000001116b5610 __swift_memcpy195_8 + 14237
	29  Samples                             0x0000000107d13f3e $s7Samples0A3AppV5$mainyyFZ + 30
	30  Samples                             0x0000000107d14049 main + 9
	31  dyld                                0x0000000108e2a384 start_sim + 10
	32  ???                                 0x00000001178ea41f 0x0 + 4690191391
)
libc++abi: terminating due to uncaught exception of type NSException
```

</details>

**Review by yo1995 — APPROVED** [review]

File: Shared/Samples/Display points using clustering feature reduction/README.md [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+* AggregateGeoElement
+* FeatureLayer
+* FeatureReduction
+* GeoElement
+* IdentifyLayerResult
+
+## About the data
+
+This sample uses a [web map](https://www.arcgis.com/home/item.html?id=8916d50c44c746c1aafae001552bad23) that displays the [Esri Global Power Plants](https://www.arcgis.com/home/item.html?id=eb54b44c65b846cca12914b87b315169) feature layer with feature reduction enabled. When enabled, the aggregate features symbology shows the color of the most common power plant type, and a size relative to the average plant capacity of the cluster.
```
Comment (pgruenler) [inline Shared/Samples/Display points using clustering feature reduction/README.md:35]:
```suggestion
This sample uses a [web map](https://www.arcgis.com/home/item.html?id=8916d50c44c746c1aafae001552bad23) that displays the [Esri Global Power Plants](https://www.arcgis.com/home/item.html?id=eb54b44c65b846cca12914b87b315169) feature layer with feature reduction enabled. When enabled, the cluster's symbology shows the color of the most common power plant type, and a size relative to the average plant capacity of the cluster.
```
The perspective of this suggestion is that the result of aggregating features is a cluster and that cluster is symbolized. If that's not the right way to think of it then don't us this suggestion. If you don't use it then "features" should probably be possessive "feature's"

**Review by pgruenler — COMMENTED** [review]
Looks good. Just one nit picky comment.

**Comment by yo1995** [comment]
I'll add Sarat to verify if he can repro [this](https://github.com/Esri/arcgis-maps-sdk-swift-samples/pull/238#issuecomment-1671757413) issue as well.

**Review by yo1995 — APPROVED** [review]

**Review by pgruenler — APPROVED** [review]

**Review by saratchandrakarumuri — APPROVED** [review]
I tried multiple times on both iOS 15 and 16 iPads and couldn't repro the crash on my end too

**Pull request merged**